File: C++/739.每日温度.cpp
/*
 * @lc app=leetcode.cn id=739 lang=cpp
 *
 * [739] 每日温度
 */

// @lc code=start
class Solution
{
public:
    vector<int> dailyTemperatures(vector<int> &temperatures)
    {
        // 单调栈
        int len = temperatures.size();
        vector<int> ans(len);
        for (int i = 0; i < len; i++)
        {
            while (!s.empty() && temperatures[i] > temperatures[s.top()])
            {
                int prev = s.top();
                ans[prev] = i - prev;
                s.pop();
            }
            s.push(i);
        }
        return ans;
    }

private:
    stack<int> s;
};
// @lc code=end

File: JavaScript/739.每日温度.js
/*
 * @lc app=leetcode.cn id=739 lang=javascript
 *
 * [739] 每日温度
 */

// @lc code=start
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function (temperatures) {
  const len = temperatures.length;
  const ans = new Array(len).fill(0);
  const stack = [];

  for (let i = 0; i < len; i++) {
    while (
      stack.length &&
      temperatures[i] > temperatures[stack[stack.length - 1]]
    ) {
      const prev = stack.pop();
      ans[prev] = i - prev;
    }
    stack.push(i);
  }

  return ans;
};
// @lc code=end
